src/libclipper/src/rpc_service.cpp

```diff
@@ -140,16 +145,61 @@ void RPCService::manage_service(const string address) {
       // TODO: Balance message sending and receiving fairly
       // Note: We only receive one message per event loop iteration
       log_info(LOGGING_TAG_RPC, "Found message to receive");
-
       receive_message(socket, connections, connections_containers_map,
                       zmq_connection_id, redis_connection);
     }
+    auto current_time = std::chrono::system_clock::now();
+    if (std::chrono::duration_cast<std::chrono::milliseconds>(
+            current_time - last_activity_check_time_)
+            .count() > CONTAINER_EXISTENCE_CHECK_FREQUENCY_MILLS) {
+      check_container_activity(connections_containers_map);
+      last_activity_check_time_ = current_time;
+    }
     // Note: We send all queued messages per event loop iteration
     send_messages(socket, connections);
   }
   shutdown_service(socket);
 }
 
+void RPCService::check_container_activity(
+  std::unordered_map<std::vector<uint8_t>, std::pair<VersionedModelId, int>, 
+        std::function<size_t(const std::vector<uint8_t> &vec)>>
+        &connections_containers_map) {
+
+  std::map<const vector<uint8_t>,
+           std::chrono::system_clock::time_point>::iterator it;
+  std::chrono::system_clock::time_point current_time;
+  for (it = receiving_history_.begin(); it != receiving_history_.end(); it++) {
+    current_time = std::chrono::system_clock::now();
+    if (std::chrono::duration_cast<std::chrono::milliseconds>(current_time -
+                                                              it->second)
+            .count() > CONTAINER_ACTIVITY_TIMEOUT_MILLS) {
+      /** if the amount of time that has elapsed between the current time and the time of last
+      receiving from the container is greater than the threshold, then we want to 
+      call the inactive_container_callback_ */
+
+      const vector<uint8_t> connection_id = it->first;
+      auto container_info_entry =
+            connections_containers_map.find(connection_id);
+        if (container_info_entry == connections_containers_map.end()) {
+          throw std::runtime_error(
+              "Failed to find container that was previously registered via "
+              "RPC");
+        }
+        std::pair<VersionedModelId, int> container_info =
+            container_info_entry->second;
+
+        VersionedModelId vm = container_info.first;
+        int replica_id = container_info.second;
+      TaskExecutionThreadPool::submit_job(
+            vm, replica_id, inactive_container_callback_, vm, replica_id);
+      
+      log_info(LOGGING_TAG_RPC, "lost contact with a container");
+      receiving_history_.erase(it);
```

Comment:
@Corey-Zumar after we exceed  CONTAINER_ACTIVITY_TIMEOUT_MILLIS, we enter this if statement. The log statement on line 197 is logged correctly. However, `submit_job` for the `inactive_container_callback` is not being executed unless a request is sent. But, these 2 statements are right after each other, so I'm not really sure why the `inactive_container_callback_` is only executed when a request is sent.

Reply:
The issue here is that the `TaskExecutionThreadPool` has a single thread for each container, and the callbacks get executed sequentially on that threadpool. The `inactive_container_callback_` is blocked by a call to `on_container_ready` which is already in the threadpool queue. `on_container_ready` blocks until there is a request to dequeue here (`get_batch()` is a blocking call). https://github.com/ucbrise/clipper/blob/42abb257976f8384f96febab270c8fd0baafb906/src/libclipper/include/clipper/task_executor.hpp#L449-L450

I think the easiest fix is to make `get_batch()` a non-blocking call by changing the call from `wait` to `wait_for` on line 144, and waiting for 100ms: https://github.com/ucbrise/clipper/blob/42abb257976f8384f96febab270c8fd0baafb906/src/libclipper/include/clipper/task_executor.hpp#L144

Instead of blocking until there's a new batch available, we will just return a batch of size 0. Then, in the else statement on `on_container_ready`, instead of logging an error, submit `on_container_ready` to the threadpool again:
https://github.com/ucbrise/clipper/blob/42abb257976f8384f96febab270c8fd0baafb906/src/libclipper/include/clipper/task_executor.hpp#L479-L485
If no other callbacks get submitted to the threadpool, this will be functionally the same as making `get_batch` blocking. But it will allow the `inactive_container_callback` to be executed without having to wait for another request to come in before the container is removed.

Reply:
@Corey-Zumar what do you think of my suggested modification?

Reply:
@dcrankshaw This sounds similar to our previous approach (circa last April) of resubmitting the `on_container_ready` callback to the threadpool if a query was not instantaneously available; we did not wait for 100ms. That approach seemed to be CPU-intensive, but I would imagine that the 100 millisecond wait would alleviate some of the overhead. 

Another option would be to create a garbage collection service backed by a single thread that handles this cleanup without relying upon threadpool execution. This service could later be extended to perform other GC-related tasks as they arise. Let me know what your thoughts are.

Reply:
Yeah I was also thinking about a GC thread. A GC thread is probably the right way to do this, as you rightly point out that we are likely to want to do more cleanup in the future. We need to be careful about how we clean up though, because in `on_container_ready` we acquire a shared_ptr to the `ModelContainer` object. If that container becomes inactive, the next time a request comes in, it will be dequeued from the model queue to be sent to the inactive container.

To do proper GC, we need the model queue to not dequeue a batch to an inactive container. Maybe we could pass the container pointer to `get_batch()` explicitly instead of capturing it with closure capture, and then the model queue will check that the container it is returning the batch to is still active before returning. If it finds that the container is inactive, it can return an empty batch.

I have another question too. Is the container's threadpool removed when the container is found to be inactive? If not, it should be.